edgeless_node/src/resources/ollama.rs

```diff
@@ -0,0 +1,344 @@
+
+        let _handle = tokio::spawn(async move {
+            let mut targets = std::collections::HashMap::new();
+            while let Some(command) = receiver.next().await {
+                match command {
+                    OllamaCommand::Chat(model_name, history_id, prompt, resource_id, reply_receiver) => {
+                        if let Some(target) = targets.get(&resource_id) {
+                            let result = ollama
+                                .send_chat_messages_with_history(
+                                    ollama_rs::generation::chat::request::ChatMessageRequest::new(
+                                        model_name.clone(),
+                                        vec![ollama_rs::generation::chat::ChatMessage::user(prompt)],
+                                    ),
+                                    history_id.clone(),
+                                )
+                                .await;
+                            let response = match result {
+                                Ok(res) => Ok((*target, res.message.unwrap().content)),
+                                Err(err) => anyhow::Result::Err(anyhow::anyhow!(
+                                    "Ollama error with model {}, history_id {}: {}",
+                                    model_name,
+                                    history_id,
+                                    err
+                                )),
+                            };
+                            let _ = reply_receiver.send(response);
+                        }
+                    }
+                    OllamaCommand::Patch(resource_id, target) => {
+                        targets.insert(resource_id, target);
+                    }
+                };
+            }
+        });
+        Self {
+            inner: std::sync::Arc::new(tokio::sync::Mutex::new(OllamaResourceProviderInner {
+                resource_provider_id,
+                dataplane_provider,
+                instances: std::collections::HashMap::new(),
+                sender,
+                _handle,
+            })),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl edgeless_api::resource_configuration::ResourceConfigurationAPI<edgeless_api::function_instance::InstanceId> for OllamaResourceProvider {
+    async fn start(
+        &mut self,
+        instance_specification: edgeless_api::resource_configuration::ResourceInstanceSpecification,
+    ) -> anyhow::Result<edgeless_api::common::StartComponentResponse<edgeless_api::function_instance::InstanceId>> {
+        let mut lck = self.inner.lock().await;
+        let new_id = edgeless_api::function_instance::InstanceId::new(lck.resource_provider_id.node_id);
+        let dataplane_handle = lck.dataplane_provider.get_handle_for(new_id.clone()).await;
+
+        // Read configuration
+        let model = match instance_specification.configuration.get("model") {
```

Comment:
We may be able to configure the resource with a list of available models and discard impossible resource requests.

Reply:
The set of available models may change because new models can be added at run-time to the ollama server.

We could check if the model requested is supported _as of now_ and discard if it is not, though.

However, resource creation failure is something that, in general in EDGELESS, is still a bit fuzzy. So, I would postpone this to a later time.